File: golang/add_two_numbers.go
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	// setup dummy head ptr
	head := &ListNode{
		Val: 0,
	}
	current, carry := head, 0

	// determine sum and carry, create node, attach to list
	for l1 != nil && l2 != nil {
		val := (l1.Val + l2.Val + carry) % 10
		carry = (l1.Val + l2.Val + carry) / 10
		node := &ListNode{
			Val: val,
		}
		current.Next = node
		current = node

		l1 = l1.Next
		l2 = l2.Next
	}

	// fill in for remaining list
	for l1 != nil {
		val := (l1.Val + carry) % 10
		carry = (l1.Val + carry) / 10
		node := &ListNode{
			Val: val,
		}
		current.Next = node
		current = node

		l1 = l1.Next
	}
	for l2 != nil {
		val := (l2.Val + carry) % 10
		carry = (l2.Val + carry) / 10
		node := &ListNode{
			Val: val,
		}
		current.Next = node
		current = node

		l2 = l2.Next
	}

	// if overflow carry
	if carry == 1 {
		node := &ListNode{
			Val: 1,
		}
		current.Next = node
		current = node
	}

	return head.Next
}

File: typescript/add_two_numbers.ts
function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
	let psuedo = new ListNode(0, null)
	let current = psuedo
	let carry = 0

	while (l1 != null && l2 != null) {
		let digit = (l1.val + l2.val + carry) % 10
		carry = Math.floor((l1.val + l2.val + carry) / 10)

		current.next = new ListNode(digit, null)
		current = current.next
		l1 = l1.next
		l2 = l2.next
	}

	while (l1 != null) {
		let digit = (l1.val + carry) % 10
		carry = Math.floor((l1.val + carry) / 10)

		current.next = new ListNode(digit, null)
		current = current.next
		l1 = l1.next
	}

	while (l2 != null) {
		let digit = (l2.val + carry) % 10
		carry = Math.floor((l2.val + carry) / 10)

		current.next = new ListNode(digit, null)
		current = current.next
		l2 = l2.next
	}

	if (carry == 1) {
		current.next = new ListNode(1, null)
	}

	return psuedo.next
}
